decoder/fort.go

```diff
@@ -2,12 +2,187 @@ package decoder
+)
+
+func InitWebHookFortFromGym(gym *Gym) (fort FortWebhook) {
+	if gym == nil {
+		return
+	}
+	fort.Type = GYM.String()
+	fort.Name = gym.Name.Ptr()
+	fort.ImageUrl = gym.Url.Ptr()
+	fort.Description = gym.Description.Ptr()
+	fort.Location = Location{Latitude: gym.Lat, Longitude: gym.Lon}
+	return
+}
+
+func InitWebHookFortFromPokestop(stop *Pokestop) (fort FortWebhook) {
+	if stop == nil {
+		return
+	}
+	fort.Type = POKESTOP.String()
+	fort.Name = stop.Name.Ptr()
+	fort.ImageUrl = stop.Url.Ptr()
+	fort.Description = stop.Description.Ptr()
+	fort.Location = Location{Latitude: stop.Lat, Longitude: stop.Lon}
+	return
+}
+
+func CreateFortWebhooks(ctx context.Context, dbDetails db.DbDetails, ids []string, fortType FortType, change FortChange) {
+	var gyms []Gym
+	var stops []Pokestop
+	if fortType == GYM {
+		for _, id := range ids {
+			gym, err := getGymRecord(ctx, dbDetails, id)
+			if err != nil {
+				continue
+			}
+			if gym == nil {
+				continue
+			}
+			gyms = append(gyms, *gym)
+		}
+	}
+	if fortType == POKESTOP {
+		for _, id := range ids {
+			stop, err := getPokestopRecord(ctx, dbDetails, id)
+			if err != nil {
+				continue
+			}
+			if stop == nil {
+				continue
+			}
+			stops = append(stops, *stop)
+		}
+	}
+	for _, gym := range gyms {
+		fort := InitWebHookFortFromGym(&gym)
+		CreateFortWebHooks(&fort, &FortWebhook{}, change)
+	}
+	for _, stop := range stops {
+		fort := InitWebHookFortFromPokestop(&stop)
+		CreateFortWebHooks(&fort, &FortWebhook{}, change)
+	}
+}
+
+func CreateFortWebHooks(old *FortWebhook, new *FortWebhook, change FortChange) {
+	if change == NEW {
+		areas := geo.MatchGeofences(statsFeatureCollection, new.Location.Latitude, new.Location.Longitude)
+		hook := map[string]interface{}{
+			"change_type": change.String(),
+			"new": func() interface{} {
+				bytes, err := json.Marshal(new)
+				if err != nil {
+					return nil
+				}
+				return json.RawMessage(bytes)
+			},
+		}
+		webhooks.AddMessage(webhooks.FortUpdate, hook, areas)
+	} else if change == REMOVAL {
+		areas := geo.MatchGeofences(statsFeatureCollection, old.Location.Latitude, old.Location.Longitude)
+		hook := map[string]interface{}{
+			"change_type": change.String(),
+			"old": func() interface{} {
+				bytes, err := json.Marshal(old)
+				if err != nil {
+					return nil
+				}
+				return json.RawMessage(bytes)
+			},
+		}
+		webhooks.AddMessage(webhooks.FortUpdate, hook, areas)
+	} else if change == EDIT {
+		areas := geo.MatchGeofences(statsFeatureCollection, new.Location.Latitude, new.Location.Longitude)
+		var editTypes []string
+		if old.Name != new.Name {
+			editTypes = append(editTypes, "name")
+		}
+		if old.Description != new.Description {
+			editTypes = append(editTypes, "description")
+		}
+		if old.ImageUrl != new.ImageUrl {
+			editTypes = append(editTypes, "image_url")
+		}
+		if old.Location != new.Location {
+			editTypes = append(editTypes, "location")
+		}
+		hook := map[string]interface{}{
```

Comment:
```golang
hook := map[string]interface{}{
			"change_type": change.String(),
			"edit_types":  editTypes,
			"old": old,
			"new": new,
}
```

is it possible to use it like that? or do we need the json.RawMessage?! :)

Reply:
This works, except old and new would always be present since there is no ability to do a 'omitempty'
If you did it as a structure you could tag old new and edit_types with omitempty
Or you can create a map as you have and just add old and new in if non nil
Except if you want old and new to always be present (ie with a null value) then this work work I would guess

Reply:
In that case I already know that there are differences and old and new are never empty :)